### PR TITLE
simplify the test workflow

### PR DESCRIPTION
**Pull request by emmaai**
### Proposed changes
- Simplify the test workflow and make it reusable in `dea-sandbox`
- Remove `index` container, install the indexing tool in `sandbox` instead
- Make bash script pass the failure

### Closes issues (optional)
N/A
### Checklist (replace `[ ]` with `[x]` to check off)
- [ ] Notebook created using the [DEA-notebooks template](https://github.com/GeoscienceAustralia/dea-notebooks/tree/develop)
- [ ] Remove any unused Python packages from `Load packages`
- [ ] Remove any unused/empty code cells
- [ ] Remove any guidance cells (e.g. `General advice`)
- [ ] Ensure that all code cells follow the [PEP8 standard](https://www.python.org/dev/peps/pep-0008/) for code. The `jupyterlab_code_formatter` tool can be used to format code cells to a consistent style: select each code cell, then click `Edit` and then one of the `Apply X Formatter` options (`YAPF` or `Black` are recommended).
- [ ] Include relevant tags in the final notebook cell (refer to the [DEA Tags Index](https://docs.dea.ga.gov.au/genindex.html), and re-use tags if possible)
- [ ] Clear all outputs, run notebook from start to finish, and save the notebook in the state where all cells have been sequentially evaluated
- [ ] Test notebook on both the `NCI` and `DEA Sandbox` (flag if not working as part of PR and ask for help to solve if needed)
- [ ] If applicable, update the `Notebook currently compatible with the NCI|DEA Sandbox environment only` line below the notebook title to reflect the environments the notebook is compatible with

### CODE REVIEW

**Comment by robbibt** [comment]
Hey @emmaai! Is the test failure here expected? Or do you think it will start to work once it's merged?

**Comment by emmaai** [comment]
no expected, should pass once able to pull the new sandbox image from ecr.

File: .github/workflows/test_notebooks.yml
Comment (robbibt) [inline .github/workflows/test_notebooks.yml]: What is JWT? Can we update the comment to clarify this?

File: docker-compose.yml
Comment (robbibt) [inline docker-compose.yml:17]: I think this needs to point to the specific Sandbox image that is used on the "Default environment" on Prod Sandbox. E.g. `2.0.0`?

(We're going to have to update this every time we change the Prod Sandbox image, which isn't ideal - I'd love us to have a "stable" tag that we could use to set both the prod Sandbox image and this line too!)

File: Tests/setup_test_datacube.sh
Comment (robbibt) [inline Tests/setup_test_datacube.sh:5]: What do these lines do? Can we add a code comment to make it clear for future readers?

File: .github/workflows/test_notebooks.yml
Comment (robbibt) [inline .github/workflows/test_notebooks.yml]: Do we need to run `docker-compose down` here?

File: Tests/test_notebooks.sh
Comment (robbibt) [inline Tests/test_notebooks.sh:5]: Same question as above

**Review by robbibt — CHANGES_REQUESTED** [review]
Hi @emmaai, this looks really nice - I really like splitting out the tests into a sh script, and lots of great simplifications of the workflow!

I've added a few comments, mainly around some extra clarifications of some of the changes, and a likely change in Sandbox image to use.

Are the failed tests here OK? Do we need them to pass before we merge?

File: docker-compose.yml
Comment (emmaai) [reply docker-compose.yml:17]: The whole point of setting up automated test workflow is to guarantee sandbox image not breaking notebooks in the update, or for us to be notified asap if it does require changes not back compatible. Any release will be based the current `latest` for docker image.  Hence if the test is performed on `latest`, it should be working on any image in the life cycle of `latest -> release -> latest`.

File: .github/workflows/test_notebooks.yml
Comment (emmaai) [reply .github/workflows/test_notebooks.yml]: Everything will be killed when the workflow reaches the end.

File: .github/workflows/test_notebooks.yml
Comment (emmaai) [reply .github/workflows/test_notebooks.yml]: JWT=json web token, as indicated by the name `id-token`. Thought it was a well-known term for AWS auth. (or was it not?)

Comment (emmaai) [reply .github/workflows/test_notebooks.yml:11]: JWT=json web token, as indicated by the name `id-token`. Thought it was a well-known term for AWS auth. (or was it not?)

File: docker-compose.yml
Comment (robbibt) [reply docker-compose.yml:17]: That's only one of the two use cases for our tests - they fullfill two different but important purposes on DEA Notebooks vs DEA Sandbox:

1) On DEA Sandbox: to guarantee changes to the Sandbox image don't break notebooks. For this, we need to test against `latest`.

2) On DEA Notebooks: to guarantee changes to Jupyter Notebooks and Python code don't break other code in the DEA Notebooks repo. For this, we need to test against **the same Sandbox image that is used for the "Default environment" on Prod Sandbox.** 

The #2 is really important for us - and testing against `latest` is useless for that because `latest` isn't available to general public users on the DEA Sandbox. We definitely need to run our tests against something like `2.0.0` for DEA Notebook's tests.

File: .github/workflows/test_notebooks.yml
Comment (robbibt) [reply .github/workflows/test_notebooks.yml]: Thanks Emma - most DEA Notebooks maintainers (including myself) won't be familiar with AWS terminology. If we can expand this out to "JSON web token" in the code comment that would be great.

**Comment by emmaai** [comment]
Ok, how about this, I’ll create a tag named `stable` for every release in
sandbox image, which will be the default on jupyterhub, for most of time
unless we need to roll back when broken. Then we don’t need to change the
tag in the workflow yaml any more.

**Comment by robbibt** [comment]
> @emmaai Ok, how about this, I’ll create a tag named `stable` for every release in sandbox image, which will be the default on jupyterhub, for most of time unless we need to roll back when broken. Then we don’t need to change the tag in the workflow yaml any more.

I think that would be fantastic!

File: .github/workflows/test_notebooks.yml
Comment (robbibt) [inline .github/workflows/test_notebooks.yml]:
```suggestion
  id-token: write   # This is required for requesting the JSON web token
```

File: docker-compose.yml
Comment (emmaai) [reply docker-compose.yml:17]: Can we merge as it is though, for the time being? I have some changes to make to `dea-sandbox` after merging this pr. I'll add new tag together.  Then I can come back again and change here when the image is ready to pull.

Comment (robbibt) [reply docker-compose.yml:17]: Yep sounds good, will mark as approved now

**Review by robbibt — APPROVED** [review]
Looks great, happy to merge and then update the image tag later. Thanks Emma!

**Pull request merged**
